framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,691 @@
+     * <p>Each annotation is followed by a space, to separate it from following Java code.
+     *
+     * @param annos the annotations to format
+     * @return all annotations in {@code annos}, each followed by a space, in a form suitable to be
+     *     printed as Java source code
+     */
+    private static String formatAnnotations(Collection<? extends Annotation> annos) {
+        StringBuilder sb = new StringBuilder();
+        for (Annotation tla : annos) {
+            if (!isInternalJDKAnnotation(tla.def.name)) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the type of an array so that it is printable in Java source code, with the
+     * annotations from the scenelib representation added in appropriate places.
+     *
+     * @param scenelibRep the array's scenelib type element
+     * @param javacRep the representation of the array's type used by javac
+     * @return the type formatted to be written to Java source code, followed by a space character
+     */
+    private static String formatArrayType(ATypeElement scenelibRep, ArrayType javacRep) {
+        List<ATypeElement> scenelibRepInJavacOrder = getSceneLibRepInJavacOrder(scenelibRep);
+        return formatArrayTypeImpl(scenelibRepInJavacOrder, javacRep);
+    }
+
+    /**
+     * This method returns each array level in scenelib's representation of an array in a list in
+     * the same order used by javac. This is necessary because javac's TypeMirror and its
+     * derivatives represent arrays differently than scene-lib does.
+     *
+     * <p>If we label an array as such: (0) int (1) [] (2) [] (3) [], then level 0 is the component
+     * type, level 1 is the "outermost" type, and 3 is the "innermost" array type. Scene-lib's
+     * representation of this type is a nested ATypeElement, with this structure: (1) - (2) - (3) -
+     * (0). The TypeMirror, on the other hand, represents the type like this: (3) - (2) - (1) - (0),
+     * for ease of printing. This method therefore descends through the scenelib structure until it
+     * finds the component, adding each item to a list. It then reverses the list, and then adds the
+     * component type to the end.
+     *
+     * <p><a
+     * href="https://checkerframework.org/jsr308/specification/java-annotation-design.html#array-syntax">The
+     * JSR 308 specification</a> explains the reasoning for scenelib's representation.
+     *
+     * @param scenelibRep scenelib's representation of an array type
+     * @return a list of the array levels in scenelib's representation, but in the order used by
+     *     javac
+     */
+    private static List<ATypeElement> getSceneLibRepInJavacOrder(ATypeElement scenelibRep) {
+        List<ATypeElement> result = new ArrayList<>();
+        ATypeElement array = scenelibRep;
+        ATypeElement component = getNextArrayLevel(scenelibRep);
+        do {
```

Comment:
I'm concerned that this loop may misbehave.  As noted in the documentation of `getNextArrayLevel`, scene-lib might not contain all information about all levels of the array.  Suppose that scene-lib contains (1) (2) (3) but not the final (0).  Then it seems this routine will return in in the order (2) (1) (3), which is incorrect.

I see two ways to fix this:
 * use scene-lib's order instead of javac's order as the canonical one used in `SceneToStubWriter`.
 * have this method return a list that may contain nulls where scene-lib has no information.  This method would always return a full-length list even if repeated calls to `getNextArrayLevel` wouldn't give all the information about the type.